### DISCUSSION TITLE
Lets brainstorm what this project is going to be!!

### DISCUSSION THREAD

**Discussion by satyajitnayk** [satyajitnayk/order-management#1] [Ideas]
**Fastify can handle comparatively higher no of request/sec**
![image](https://github.com/satyajitnayk/order-management/assets/32722867/1f47be51-cb46-4711-99eb-1b95e1308fff)

Following are few ideas we can build
- [x] shipping order service - we have a paid order ,need to create shipment for user

@JyotiRanjanSatapathy  you can add more with little descriptions too!

**Comment by satyajitnayk** [comment]
### Project Idea: Order Management Queue Service Using Fastify Framework

#### Overview
Develop an order management queue service designed for high performance and efficiency, leveraging the Fastify framework. This service will handle order processing in a queue-based system, ensuring organized and scalable management of customer orders.

#### Tech Stack
- **Backend Framework**: Fastify for fast and low overhead HTTP requests.
- **Queue Management**: Redis or RabbitMQ for managing the order queue.
- **Database**: MongoDB or PostgreSQL for storing order data.
- **Authentication**: JWT (JSON Web Tokens) for secure API endpoints.
- **Testing**: Jest or Mocha for backend testing.
- **Deployment**: Docker for containerization and Kubernetes for orchestration.

#### Key Features
1. **Order Submission Endpoint**: Allows submission of new orders, which are then placed in a queue.
2. **Queue Processing System**: Processes orders in the queue efficiently, maintaining order and handling concurrency.
3. **Real-Time Order Tracking**: Provides real-time updates on order status for customers and admins.
4. **Admin Dashboard**: For monitoring and managing the order queue.
5. **Scalability**: Efficient handling of high-volume order traffic.
6. **API Integration**: For integration with other systems like payment gateways or inventory management.
7. **Error Handling and Logging**: Robust error handling mechanisms and detailed logging for troubleshooting.

#### Development Phases
1. **Planning & Design**: Define the system architecture, including the queue mechanism and database schema.
2. **Setup Fastify Environment**: Initialize Fastify project and set up necessary plugins.
3. **Queue System Integration**: Integrate and configure Redis or RabbitMQ for order queuing.
4. **API Development**: Develop RESTful APIs for order submission and tracking.
5. **Admin Interface**: Create a simple interface for administrators to monitor and manage the queue.
6. **Testing**: Conduct thorough tests for functionality, performance, and security.
7. **Deployment**: Deploy using Docker and Kubernetes for scalability and manageability.

#### Potential Challenges
- Handling high concurrency and maintaining performance under load.
- Ensuring data consistency and integrity throughout the order processing lifecycle.
- Implementing an effective real-time update system for order tracking.

**Comment by JyotiRanjanSatapathy** [comment]
Requirements

1.We have orderAPI which validates an order for various products and sends a response of paymentGatewayUrl along with redirectUrl.
2.redirectUrl is nothing but an internal service to check payment status fulfillment which takes txnId and orderId to verfiy the request
3.After a successful verification , the idea is to push the successful order into a queue.
4.The next process is to create a shipmentOrder ,then create a pickupShipmentRequest and update the details subsequently on the required tables i.e order where shipmentOrderId along with pickupShipmentRequestId. Such details we will get are shipmentOrderId,shipmentId,AWB
5.Once the product is shipped , a notification shall be generated to the user regarding your order has been picked.

**Reply by satyajitnayk** [reply]
Tried to elaborate what you discussed above 🙂🚀. Let me know if something needs to be updated @JyotiRanjanSatapathy !
### Order Processing Workflow

This section outlines the workflow of our order processing system, designed to ensure seamless transactions and efficient handling of product orders.

- [x] Order Validation and Payment Initiation
- **OrderAPI**: This API plays a critical role in validating orders for various products. Upon successful validation, it generates a `paymentGatewayUrl` for processing payments.
- **Redirect URL**: Accompanying the payment URL, a `redirectUrl` is provided. This URL is an internal service endpoint used to verify payment status. It requires `txnId` (transaction ID) and `orderId` to authenticate and process the request.

- [ ] Verification and Queueing
- After successful payment verification, the validated order is queued for the next processing stage. This ensures that orders are processed efficiently and in sequence.

- [ ] Shipment Order Creation and Tracking
- **Shipment Order**: Upon dequeuing, a `shipmentOrder` is created. This step is crucial for preparing the order for shipment.
- **Pickup Shipment Request**: Subsequently, a `pickupShipmentRequest` is generated to handle the physical aspect of the order processing.
- **Database Updates**: Key details such as `shipmentOrderId` and `pickupShipmentRequestId` are updated in the relevant database tables, specifically in the `order` table. Additional details like `shipmentOrderId`, `shipmentId`, and `AWB` (Air Waybill) are also recorded.

- [ ] Post-Shipment Notifications
- Once the product is dispatched and on its way to the customer, a notification is generated and sent to the user. This notification confirms that the order has been shipped, providing peace of mind and enhancing the customer experience.

This workflow ensures a smooth transition from order placement to shipment, maintaining high standards of efficiency and customer satisfaction.

**Comment by satyajitnayk** [comment]
Let's discuss about the system & DB structure & tradeoffs @JyotiRanjanSatapathy

**Reply by JyotiRanjanSatapathy** [reply]
@satyajitnayk 
Exploration Notes:

1. **Fastify Framework:**
   - **Objective:** Explore the Fastify framework for Node.js.
   - **Key Points:**
     - Fastify is a web framework focused on providing high performance.
     - It is designed to be efficient, lightweight, and developer-friendly.
     - Investigate its key features, advantages, and use cases.
     - Check for community support, plugins, and integrations.

2. **Database Structure Discussion:**
   - **Objective:** Discuss and refine the database structure for the project.
   - **Key Points:**
     - Identify and define the core entities and relationships within the database.
     - Discuss data models, associations, and any specific requirements.
     - Consider scalability, data integrity, and ease of querying.

3. **ShipRocket Shipping Service:**
   - **Objective:** Explore ShipRocket as the chosen third-party shipping service.
   - **Key Points:**
     - Investigate ShipRocket's features and capabilities.
     - Understand integration options and APIs for connecting with the service.
     - Check for documentation on shipping workflows, label generation, and order tracking.
     - Evaluate pricing, support, and any limitations.

These explorations will provide valuable insights into the technology stack, database architecture, and the chosen shipping service, setting the foundation for informed decision-making in the project development.

**Reply by satyajitnayk** [reply]
@JyotiRanjanSatapathy did you get some time to work on anything here?

**Reply by satyajitnayk** [reply]
@JyotiRanjanSatapathy just a reminder :)

**Reply by JyotiRanjanSatapathy** [reply]
@satyajitnayk Working on the same.Will update soon

**Reply by satyajitnayk** [reply]
@JyotiRanjanSatapathy just a reminder :)
